futures-util/src/sink/with_flat_map.rs

```diff
@@ -60,74 +76,89 @@ where
         self.sink
     }
 
-    fn try_empty_stream(&mut self, cx: &mut task::Context) -> Poll<(), S::SinkError> {
-        if let Some(x) = self.buffer.take() {
-            match self.sink.poll_ready(cx)? {
-                Async::Ready(()) => self.sink.start_send(x)?,
-                Async::Pending => {
-                    self.buffer = Some(x);
-                    return Ok(Async::Pending);
+    fn try_empty_stream(self: PinMut<Self>, cx: &mut task::Context) -> Poll<Result<(), S::SinkError>> {
+        let WithFlatMap {
+            sink,
+            f: _,
+            stream,
+            buffer,
+            _phantom,
+        } = unsafe { PinMut::get_mut_unchecked(self) };
+        let mut sink = unsafe { PinMut::new_unchecked(sink) };
+        let mut stream = unsafe { PinMut::new_unchecked(stream) };
```

Comment:
It'd be nicer if these used the methods. The unsafe code seems unnecessary.

Reply:
Also, I think temporary variables are not really needed. `self.sink()` can simply be called where it's needed

Reply:
`self.sink()` and other methods produce overlapping borrows of `self` (because each individually borrows `self` mutably, so you can't access more than one at a time this way).